# Pandas and numpy reference
 

If you're coming from DSCI 101 or 102, you've been working with very specific data objects (the `datascience` **table**). We've been using these tables for teaching purposes, but the real world of Python for data science uses the **DataFrame**, or **df**, from the `pandas` package. They are, for all intents and purposes, identical to our tables. However, we can apply a wider array of functions and methods to dfs, and we work with them in slightly different ways. Methods we used in DSCI 101 and 102 won't work on dfs. We have made this module to introduce you to this new object and transition your workflow from the `datascience.table` to the `pandas.DataFrame`. 

In [1]:
# You're familiar with numpy, but now we need to import pandas as well
import pandas as pd
import numpy as np

# You may need to change the path to connect to the data you want to use ...
path = 'https://github.com/oregon-data-science/DSCI101/raw/main/data/'

'imports complete'

'imports complete'

We'll also start using `matplotlib` for plotting instead of the `datascience` .plot and .hist methods. Again, this is a widely used framework in the real world of data science. We have a dedicated module for plotting in `matplotlib` separate from this one. 

In [2]:
# Load matplotlib 
%matplotlib inline
import matplotlib.pyplot as plt

# Apply a generic style to all plots
plt.style.use('fivethirtyeight')

### Table of Contents

1) [Using Numpy](#Section1)

2) [Creating and Displaying DataFrames](#Section2)

3) [Accessing Values and Subsetting DataFrames](#Section3)

4) [Modifying DataFrames](#Section4)

5) [Train and Test Splits](#Section5)

6) [Quick Reference](#Section6)

# 1. Using numpy
<a id='Section1'></a>

### Creating arrays

To create an array, we use `np.array`. Note that this is different from the `make_array` you used in the `datascience` package.

In [3]:
my_array = np.array([1,2,6,7])
my_array

array([1, 2, 6, 7])

We often need to instatiate empty arrays that we wish to fill out iteratively with for-loops. You could do this in `datascience` with `make_array`, but in numpy we use `np.empty(0)`.

In [4]:
my_array = np.empty(0)
my_array

array([], dtype=float64)

In [5]:
# It has a length of 0
len(my_array)

0

Use `np.append` to iteratively build an array by looping

In [6]:
for i in np.arange(10):
    my_array = np.append(my_array, i)
    
my_array

array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])

### Accessing values

#### Positional indices

To access values in numpy arrays, we use brackets `[]` to subset elements by their positional index. The equivalent process in `datascience` was the `.item()` method. 

In [7]:
my_array[2]

2.0

In [8]:
my_array[2:6]

array([2., 3., 4., 5.])

#### Random samples

The function used here is the same as in `datascience`

In [9]:
np.random.choice(my_array)

5.0

#### The percentile function

The percentile function is almost identical in numpy, only that the order in which the arguments are taken is switched: **the first argument should be your array, and the second your percentile**. 

In [10]:
np.percentile(my_array, 2.5)

0.225

# 2) Creating, displaying and summarizing DataFrames
<a id='Section2'></a>

## Creating and loading DataFrames

In Python, we have general data structures called dictionaries. They are objects that have a 'key' (equivalent to a column name is a table) and values associated with each key (equivalent to data in a table column). Let's create a simple dictionary and then show how it can be turned into a 'pandas' DataFrame. 

In [11]:
# Create a dictionary
data_dict = {
    'col_1': [3, 2, 1, 0],
    'col_2': ['a', 'b', 'c', 'd']
}

data_dict

{'col_1': [3, 2, 1, 0], 'col_2': ['a', 'b', 'c', 'd']}

Even though a dictionary doesn't look like a table, it can act like one. This type of notation is the basis for most forms of data objects, table or otherwise. Accordingly, we can easily convert dictionaries to a DataFrame with `pd.DataFrame()` or `from_dict()`. 

Note that the dictionary **needs to be rectangular** for this to work.

In [12]:
# Create a tiny test DataFrame
data_df = pd.DataFrame(data_dict)

data_df.head()  # The bold numbers on the left are row index values, dictionary keys become table column names

col_1 col_2
0      3     a
1      2     b
2      1     c
3      0     d

### Read a <font color='blue'>.csv file</font> into a pandas <font color='blue'>DataFrame</font>

While it's good to know how to create DataFrames manually and there are occasional uses for this skill, you will most commonly load data with Python from external sources. This is functionally identical to `read_table()`. As a reminder, .csv stands for **c**omma **s**eparated **v**alues, and is a very common format for storing tabular data as a file. 

### <font color='blue'>pd.read_csv(</font><font color='gray'>path_to_file</font><font color='blue'>)</font>

In [13]:
# Read a DataFrame (similar to a table) from a .csv file
sky_df = pd.read_csv(path + 'skyscrapers_v2.csv')

# Display the first 9 rows of data along with column names
sky_df.head(9)

name         material           city  \
0             One World Trade Center  mixed/composite  New York City   
1                       Willis Tower            steel        Chicago   
2                    432 Park Avenue         concrete  New York City   
3  Trump International Hotel & Tower         concrete        Chicago   
4              Empire State Building            steel  New York City   
5              Bank of America Tower  mixed/composite  New York City   
6                 Stratosphere Tower         concrete      Las Vegas   
7                         Aon Center            steel        Chicago   
8                John Hancock Center            steel        Chicago   

       height  completed  
0  541.299988       2014  
1  442.140015       1974  
2  425.500000       2015  
3  423.220001       2009  
4  381.000000       1931  
5  365.799988       2009  
6  350.220001       1996  
7  346.260010       1973  
8  343.690002       1969

As you can see, `.head()` is similar to `.show()` from the datascience package. Other tools for getting a quick glance at your data include `.tail()` (shows the last rows of data) and `.sample()` (you know what sample does). 

## Displaying your DataFrame

Most datasets we work with will be far too large to look at the entire thing. The following methods will let us look at small parts of the DataFrame to develop a sense of the shape, structure and properties of our data. 

### <font color='gray'>df</font><font color='blue'>.head(</font><font color='gray'></font><font color='blue'>)</font> and <font color='gray'>df</font><font color='blue'>.tail(</font><font color='gray'></font><font color='blue'>)</font> Display beginning or end of your DataFrame
`
df.head(5) # Show 1st 5 rows
df.tail(5)  # Show last 5 rows
df.sample(n=3) # Show random 3 rows
df.sample(frac=0.15 # Show random 0.15 (15%) of rows
df.sample(n=4, random_state=311) # show same random 4 rows each time
`

In [14]:
# Display the first 5 rows of your data
sky_df.head(5)

name         material           city  \
0             One World Trade Center  mixed/composite  New York City   
1                       Willis Tower            steel        Chicago   
2                    432 Park Avenue         concrete  New York City   
3  Trump International Hotel & Tower         concrete        Chicago   
4              Empire State Building            steel  New York City   

       height  completed  
0  541.299988       2014  
1  442.140015       1974  
2  425.500000       2015  
3  423.220001       2009  
4  381.000000       1931

In [15]:
# Display the last 5 rows of your data
sky_df.tail(5)

name  material        city  \
1776                                   Meridian Plaza I  concrete  Sacramento   
1777                                Wainwright Building     steel   St. Louis   
1778                            Hilliard Senior Homes I  concrete     Chicago   
1779                           Hilliard Senior Homes II  concrete     Chicago   
1780  Grange Mutual Insurance Headquarters, South Bu...     steel    Columbus   

         height  completed  
1776  46.000000       2003  
1777  44.810001       1892  
1778  44.500000       1966  
1779  44.500000       1966  
1780  43.889999       2010

### <font color='gray'>df</font><font color='blue'>.sample(</font><font color='gray'></font><font color='blue'>)</font> Retrieve random rows of your DataFrame

This method you should be familiar with. We can sample by number of rows (**n**) or by fraction of rows (**frac**).

In [16]:
# Display 10 random rows of your data
sky_df.sample(n = 10)

name         material  \
1249                             Trump Park Avenue            steel   
919                                         Ashton         concrete   
1196                    Bay House Miami Residences         concrete   
1123                                    Viad Tower  mixed/composite   
1724                                  Cudahy Tower            steel   
1037                        45-45 Center Boulevard         concrete   
1279                       Plaza Towers Apartments         concrete   
866              Brickell on the River South Tower         concrete   
465                 Sterling Drug Company Building            steel   
1622  Texas Children's Hospital Pavilion for Women         concrete   

               city      height  completed  
1249  New York City  106.680000       1929  
919          Austin  125.500000       2009  
1196          Miami  110.000000       2015  
1123        Phoenix  114.000000       1991  
1724      Milwaukee   68.400002       1928  
1037  New York City  118.900002       2013  
1279  New York City  105.459999       1964  
866           Miami  128.929993       2007  
465   New York City  156.970001       1964  
1622        Houston   81.989998       2012

In [17]:
# Randomly display 10% of your data
sky_df.sample(frac = 0.10)

name         material           city  \
1591                    Lake Park Plaza         concrete        Chicago   
1625                   Via6 North Tower         concrete        Seattle   
1391                         The Laurel         concrete  New York City   
1412                The John Ross Tower         concrete       Portland   
1628                        Kheel Tower            steel  New York City   
...                                 ...              ...            ...   
1373                      2 Wall Street            steel  New York City   
943          Murano Grande at Portofino         concrete    Miami Beach   
64                   Wells Fargo Center  mixed/composite    Minneapolis   
1142             Western Union Building            steel  New York City   
812   Carl B. Stokes Federal Courthouse            steel      Cleveland   

          height  completed  
1591   84.750000       1969  
1625   81.400002       2013  
1391  100.000000       2009  
1412   99.059998       2007  
1628   81.379997       1926  
...          ...        ...  
1373  100.599998       1933  
943   124.059998       2003  
64    236.320007       1988  
1142  113.000000       1930  
812   131.070007       2002  

[178 rows x 5 columns]

Note than when we display subsets of our data, the total number of displayed rows is capped at 10 by default. If we displayed all 178 rows from our 10% sample, the resulting table would take up the majority of the notebook. 

### Display info about columns in your DataFrame

- `df.shape` # Returns dimensions of DataFrame
- `df.info()` # info about each column  
- `df.describe()` # statistics about each column of numbers  
- `df['column_name'].unique()` # list of all unique values found in a column  

The tools we just introduced are all great for quick checks on our data. Did the DataFrame import as I expected? Do I have the right number of columns and the right type of data in each?

We'll now show you a few ways to explore our DataFrames in more detail.

### <font color='blue'>pd.shape</font> # The dimensions of our DataFrame

`.shape` returns the number of rows and columns in our data, telling us how large our dataset is (rows * columns"), how many variables we have (columns), and how many observations we have (rows). 

In [18]:
sky_df.shape   # (number of rows, number of columns)

(1781, 5)

In [19]:
num_rows = sky_df.shape[0]
num_cols = sky_df.shape[1]

print(f'{num_rows} rows, {num_cols} columns in sky_df data frame')

1781 rows, 5 columns in sky_df data frame


Note another trick we just employed. The bracket subset in the form `object_name[index]` lets us pull out values from an array based on their location (index) in the array. `sky_df.shape[0]` means take the first value from the returned shape method. 

### <font color='blue'>df.info( )</font> # Type of each column, how many non-missing values in each column, memory use

`.info()` returns a lot of useful information about your DataFrame. This includes the class of object (DataFrame in this case), the number of rows, the columns and their names, the number of observations in each column without missing values (NULL or NA), and the type of data in each column.

It also tells you much of your computer's working memory (RAM) is occupied by the DataFrame. This is not an issue we've run into in DSCI101 and 102, but could be a problem when working with datasets w/ hundreds of thousands or millions of observations. 

In [20]:
sky_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1781 entries, 0 to 1780
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   name       1781 non-null   object 
 1   material   1781 non-null   object 
 2   city       1781 non-null   object 
 3   height     1781 non-null   float64
 4   completed  1781 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 69.7+ KB


###  <font color='blue'>df.describe( )</font> # count, mean, std, min, max of each numeric column

How about some quick statistics on all of your numeric variables?

In [21]:
sky_df.describe()

height    completed
count  1781.000000  1781.000000
mean    135.641084  1980.797305
std      49.147231    30.441838
min      43.889999  1886.000000
25%     102.410004  1967.000000
50%     128.000000  1987.000000
75%     158.800003  2007.000000
max     541.299988  2016.000000

### <font color='blue'>df['col-A'].unique( )</font> Unique values in a column

Finally, we can easily look at all unique values in a column with `.unique()`. 

In [22]:
sky_df['material'].unique()

array(['mixed/composite', 'steel', 'concrete'], dtype=object)

# 3) Accessing values and subsetting DataFrames
<a id='Section3'></a>

It's one thing to *look* at your values, but *doing* anything with them requires tools for accessing only those values we want. Below we'll show how to access specific values by

    1) Selecting rows/columns by their positional index
    
    2) Selecting columns by their name
    
    3) Subsetting rows with logical expressions

## Subsets with indices

### <font color='blue'>take some rows</font>  using row numbers (0 is first row)

We previously demonstrated the use of brackets to subset arrays. We'll do something similar again to subset DataFrames in two dimensions (rows and columns) using the `.iloc[]` method. 

In [23]:
# Return the value of the 5th row in the 1st column 
sky_one_index = sky_df.iloc[4,0]
sky_one_index

'Empire State Building'

We can also subset using multiple indices at once. An easy way to do so is with the `:` operator. It functions by returning all integers from a number on the left to the number on the right. When we include only `:` as our index, this simply means *all* values possible in the data object.   

In [24]:
# Return rows 3, 4 and 5, and return all columns
sky_via_index_numbers = sky_df.iloc[3:6, :] # Rows 3, 4, 5 but not 6, all columns :
sky_via_index_numbers

name         material           city  \
3  Trump International Hotel & Tower         concrete        Chicago   
4              Empire State Building            steel  New York City   
5              Bank of America Tower  mixed/composite  New York City   

       height  completed  
3  423.220001       2009  
4  381.000000       1931  
5  365.799988       2009

## Subsets by column name

### <font color='blue'>which columns</font> to use in a new data frame</font> --like <font color='gray'>datascience</font> <font color='darkgreen'>.select( )</font> 

As we've already discovered, it's often easier or necessary to subset out tables for just the variables we need. Conveniently, we can do this again with bracket notation. Instead of passing numeric indices, we can pass an array of column names to take. 

In [25]:
columns_to_use = ['completed', 'city', 'height']
sky2_df = sky_df[columns_to_use]
sky2_df.tail(10)

completed            city     height
1771       1910  Salt Lake City  50.000000
1772       1899   New York City  49.380001
1773       1971  Salt Lake City  49.070000
1774       1897         Atlanta  48.200001
1775       2014         Chicago  47.400002
1776       2003      Sacramento  46.000000
1777       1892       St. Louis  44.810001
1778       1966         Chicago  44.500000
1779       1966         Chicago  44.500000
1780       2010        Columbus  43.889999

### <font color='brown'>[[ ... ]]</font> is <font color='gray'>more traditional in Pandas</font> # but also <font color='darkred'>more error-prone</font>

Forgetting to use double [[ ]] is a very common error;

Use a list and then it's just my_df[my_list_of_columns] 

In [26]:
sky2_df = sky_df[['completed', 'city', 'height']]  # Without [[ ]] it won't work right!
sky2_df.tail(5)

completed        city     height
1776       2003  Sacramento  46.000000
1777       1892   St. Louis  44.810001
1778       1966     Chicago  44.500000
1779       1966     Chicago  44.500000
1780       2010    Columbus  43.889999

## Subsets with logical expressions

### <font color='blue'>Get selected rows</font>  from pandas data frame  -like <font color='gray'>datascience</font> <font color='darkgreen'>.where( )</font> 

Index and column subsets have their uses, but the ability to subset our data based on specific conditions is workhorse tool in the data science toolbox. In 101 and 102 we used `.where()` to do this, but going forward we'll use the **logical subset**. 

1) Evaluate values in a column against against conditions using logical operators, attaining a new column of True/False values.

**Logical operators** and their equivalents in the `datascience` package. 

- `==` Is equal to; `are.equal_to()`                 
- `!=` Not equal to; `are.not_equal_to()`    
- `>`  Greater than; `are.above()`           
- `>=` Greater than or equal to; `are.above_or_equal_to()`   
- `<`  Less than; `are.below()`                
- `<=` Less than or equal to; `are.below_or_equal_to()`
- `in` Contained in; `are.containg()`

In [27]:
## Get a True / False column, filtering by year completed ...
np.random.seed(1111) # to be able to repeat and get same results each time

# Evaluate every value in the 'completed' column against the logical expression 'is equal to 1979'
select_1979 = sky_df['completed'] == 1979

select_1979.sample(15)

1525    False
1725    False
236     False
479     False
306     False
452     False
1218    False
514     False
639     False
1374    False
1427    False
1247    False
1360     True
330     False
1071    False
Name: completed, dtype: bool

2) Using these True/False values, bracket subset our data returning only those rows associated with True

In [28]:
select_1979 = sky_df['completed'] == 1979

# Only put True rows into sky_1979_df
sky_1979_df = sky_df[select_1979]  

sky_1979_df.head(5)

name  material           city      height  \
201             Citigroup Center     steel    Los Angeles  190.500000   
627            45 Fremont Center     steel  San Francisco  145.000000   
645          333 Market Building     steel  San Francisco  144.000000   
929            595 Market Street  concrete  San Francisco  125.000000   
930  1100 North Lake Shore Drive  concrete        Chicago  124.970001   

     completed  
201       1979  
627       1979  
645       1979  
929       1979  
930       1979

Even better, skip the middle-object and do the whole subset in one go, nesting our logical expression in our subset brackets.

In [29]:
# You may see pandas code to get all buildings completed in 1979 using this style
sky_1979_bldgs = sky_df[sky_df['completed'] == 1979]
sky_1979_bldgs.head(5)

name  material           city      height  \
201             Citigroup Center     steel    Los Angeles  190.500000   
627            45 Fremont Center     steel  San Francisco  145.000000   
645          333 Market Building     steel  San Francisco  144.000000   
929            595 Market Street  concrete  San Francisco  125.000000   
930  1100 North Lake Shore Drive  concrete        Chicago  124.970001   

     completed  
201       1979  
627       1979  
645       1979  
929       1979  
930       1979

### Get all buildings completed <font color='blue'>before 1979</font>

In [30]:
select_before_1979 = sky_df['completed'] < 1979

# Only put True rows into sky_before_1979_df
sky_before_1979_df = sky_df[select_before_1979]  

sky_before_1979_df.head(10)

name  material           city      height  completed
1            Willis Tower     steel        Chicago  442.140015       1974
4   Empire State Building     steel  New York City  381.000000       1931
7              Aon Center     steel        Chicago  346.260010       1973
8     John Hancock Center     steel        Chicago  343.690002       1969
9       Chrysler Building     steel  New York City  318.899994       1930
21                70 Pine     steel  New York City  290.170013       1932
25     The Trump Building     steel  New York City  282.549988       1930
27       Citigroup Center     steel  New York City  278.899994       1977
35      Water Tower Place  concrete        Chicago  261.880005       1976
37             Aon Center     steel    Los Angeles  261.519989       1974

### Get all the items that are in a list using <font color='blue'>.isin(list of values)</font>

In [31]:
# Select all rows associated with the years 1899, 1904, 1921
select_dates = sky_df['completed'].isin([1899, 1904, 1921])

# select dates is an array of True / False values
sky_selected_years = sky_df[select_dates]

sky_selected_years

name material           city      height  completed
908            Crown Building    steel  New York City  126.800003       1921
1033        Park Row Building    steel  New York City  119.180000       1899
1579         6 North Michigan    steel        Chicago   85.949997       1899
1664        Santa Fe Building    steel        Chicago   78.940002       1904
1772  Bayard-Condict Building    steel  New York City   49.380001       1899

In [32]:
# Select all rows associated with the cities Atlanta, Sacramento, San Diego
select_cities = sky_df['city'].isin(['Atlanta', 'Sacramento', 'San Diego'])

# select_cities is now an array of True / False values

sky_selected_cities = sky_df[select_cities]
sky_selected_cities.sample(10)

name         material  \
1344                                The Grandview         concrete   
1148                                 Terminus 200         concrete   
66                            191 Peachtree Tower  mixed/composite   
1197  San Diego Marriott Marquis & Marina Tower 1            steel   
1087                               Sapphire Tower         concrete   
885                    Vantage Pointe Condominium         concrete   
44                            One Atlantic Center  mixed/composite   
289                                  The Atlantic         concrete   
1225                                Esquire Plaza            steel   
856                              Harbor Club East         concrete   

            city      height  completed  
1344     Atlanta  102.000000       1990  
1148     Atlanta  112.650002       2009  
66       Atlanta  234.699997       1992  
1197   San Diego  110.000000       1984  
1087   San Diego  115.800003       2008  
885    San Diego  128.020004       2009  
44       Atlanta  249.940002       1987  
289      Atlanta  175.869995       2009  
1225  Sacramento  108.199997       1999  
856    San Diego  129.240005       1992

### Show buildings (built in 1924)</font><font color='red'> & </font> (taller than 150 meters)<font color='red'>

Finally, we're not limited by singular logical expressions. We can use multiple expressions to string together filters on our data. The most common operators are **and** represented by `&` in 'pandas', and **or**, represented by `|`. When subsetting with multiple conditions, `&` means that both logical expressions must be true for a row to be returned, while `|` means that either logical expression can be true for a row to be returned. 

Note: "and" would operate on an entire column but pandas says "this does not make sense"   
Note2: & (bitwise "and") used in Pandas, works element by element   
Note3: Need parentheses ( ) & ( )

In [33]:
# ( ) parentheses needed 
select_1924_and_tall = (sky_df['completed'] == 1924) & (sky_df['height'] > 150.0)
#                      *                           *   *                        *
sky_old_tall = sky_df[select_1924_and_tall]
sky_old_tall.head()

name material           city      height  completed
315   Chicago Temple Building    steel        Chicago  173.130005       1924
451               26 Broadway    steel  New York City  158.500000       1924
487  Bank of America Building    steel      Baltimore  155.149994       1924

Alternatively, 

In [34]:
select_1924 = sky_df['completed'] == 1924
select_tall  = sky_df['height'] > 150.0

sky_old_tall3 = sky_df[select_1924 & select_tall]

sky_old_tall3.head()

name material           city      height  completed
315   Chicago Temple Building    steel        Chicago  173.130005       1924
451               26 Broadway    steel  New York City  158.500000       1924
487  Bank of America Building    steel      Baltimore  155.149994       1924

If we want either of two groups, we use `|` instead. 

### Either (completed in 1904 ) <font color='red'> | </font>  (city is Sacramento)

`select_groups = (sky_df['city'] == 'Sacramento') | (sky_df['completed'] == 1904) `

In [35]:
# Again, remember the ( )
select_city_or_date = (sky_df['city'] == 'Sacramento') | (sky_df['completed'] == 1904) 
#                     *                              *   *                           *    
sky_city_or_date = sky_df[select_city_or_date]
sky_city_or_date

name         material        city  \
965                              US Bank Tower            steel  Sacramento   
1008                    Bank of the West Tower            steel  Sacramento   
1083                                Park Tower  mixed/composite  Sacramento   
1134                         Renaissance Tower            steel  Sacramento   
1136  Joe Serna Junior California EPA Building            steel  Sacramento   
1225                             Esquire Plaza            steel  Sacramento   
1237                            Capitol Square         concrete  Sacramento   
1257       Robert T. Matsui Federal Courthouse            steel  Sacramento   
1634                             1325 J Street         concrete  Sacramento   
1635                          300 Capitol Mall            steel  Sacramento   
1664                         Santa Fe Building            steel     Chicago   
1776                          Meridian Plaza I         concrete  Sacramento   

          height  completed  
965   122.599998       2008  
1008  120.699997       2009  
1083  115.820000       1991  
1134  113.389999       1989  
1136  113.389999       2000  
1225  108.199997       1999  
1237  106.980003       1992  
1257  106.680000       1992  
1634   80.769997       1999  
1635   80.769997       1984  
1664   78.940002       1904  
1776   46.000000       2003

## Subset with 'query'

This is a good time to remind you about best practices in variable names, including column names. We should always avoid spaces in variable names (`year_completed` instead of `year completed`), as well as special characters (`tax_percent` instead of `tax%`). This is particularly important for `.query()`.

### Use df<font color='blue'>.query( )</font> to select rows

`.query()` lets us select rows using expressions like above, but with different notation. Whereas above we created arrays of True/False and used these to subset in bracket notation, `.query()` is a method that subsets for us given a logical expression. 

In [36]:
# Get all buildings completed in 1980
sky_1980 = sky_df.query('completed == 1980')
sky_1980.head(6)

name         material           city      height  completed
72     Enterprise Plaza  mixed/composite        Houston  230.399994       1980
130  Three Allen Center            steel        Houston  208.789993       1980
139       AT&T Building         concrete        Atlanta  206.350006       1980
433      767 3rd Avenue            steel  New York City  160.000000       1980
548      55 West Monroe         concrete        Chicago  150.919998       1980
603         KPMG Centre            steel         Dallas  146.610001       1980

In [37]:
# Get buildings completed before 1969
sky_pre_1969 = sky_df.query('completed < 1969')
sky_pre_1969.head(10)

name material           city      height  completed
4   Empire State Building    steel  New York City  381.000000       1931
9       Chrysler Building    steel  New York City  318.899994       1930
21                70 Pine    steel  New York City  290.170013       1932
25     The Trump Building    steel  New York City  282.549988       1930
39       Comcast Building    steel  New York City  259.079987       1933
47             28 Liberty    steel  New York City  247.809998       1961
49       MetLife Building    steel  New York City  246.279999       1963
52     Woolworth Building    steel  New York City  241.399994       1913
78       Prudential Tower    steel         Boston  228.600006       1964
89        Twenty Exchange    steel  New York City  225.860001       1931

In [38]:
# Get buildings completed in any of theses years [1899, 1904, 1921]
sky_years = sky_df.query('completed in [1899, 1904, 1921]')
sky_years.tail()

name material           city      height  completed
908            Crown Building    steel  New York City  126.800003       1921
1033        Park Row Building    steel  New York City  119.180000       1899
1579         6 North Michigan    steel        Chicago   85.949997       1899
1664        Santa Fe Building    steel        Chicago   78.940002       1904
1772  Bayard-Condict Building    steel  New York City   49.380001       1899

In [39]:
# Show buildings (built in 1924) and (taller than 150 meters)
sky_1924_or_tall = sky_df.query('(completed == 1924) and (height > 150)')
sky_1924_or_tall.head()

name material           city      height  completed
315   Chicago Temple Building    steel        Chicago  173.130005       1924
451               26 Broadway    steel  New York City  158.500000       1924
487  Bank of America Building    steel      Baltimore  155.149994       1924

In [40]:
# City Portland or height >= 400
sky_weird_or_tall = sky_df.query('(city == "Portland") or  (height >= 400)')
sky_weird_or_tall.head()

name         material           city  \
0               One World Trade Center  mixed/composite  New York City   
1                         Willis Tower            steel        Chicago   
2                      432 Park Avenue         concrete  New York City   
3    Trump International Hotel & Tower         concrete        Chicago   
377                 Wells Fargo Center            steel       Portland   

         height  completed  
0    541.299988       2014  
1    442.140015       1974  
2    425.500000       2015  
3    423.220001       2009  
377  166.419998       1973

# 4) Modifying DataFrames
<a id='Section4'></a>

We'll wrap up with a buffet of options for modifying our DataFrames, including sorting, changing values, and adding new data. 

## Sorting

In sorting by columns, we can reorganize our DataFrame in logical orders.

### <font color='blue'>.sort_values</font>(<font color='gray'>'city'</font>) # sort by data frame column</font>

If we sort by a column of string values, the DataFrame is reordered in alphabetical order of that column. 

In [41]:
sky_by_city = sky_df.sort_values('height')

sky_by_city.head(10)

name  material  \
1780  Grange Mutual Insurance Headquarters, South Bu...     steel   
1779                           Hilliard Senior Homes II  concrete   
1778                            Hilliard Senior Homes I  concrete   
1777                                Wainwright Building     steel   
1776                                   Meridian Plaza I  concrete   
1775         GEMS World Academy - Chicago, Lower School  concrete   
1774                                       FlatironCity     steel   
1773                Social and Behavioral Science Tower  concrete   
1772                            Bayard-Condict Building     steel   
1771                                    Boston Building     steel   

                city     height  completed  
1780        Columbus  43.889999       2010  
1779         Chicago  44.500000       1966  
1778         Chicago  44.500000       1966  
1777       St. Louis  44.810001       1892  
1776      Sacramento  46.000000       2003  
1775         Chicago  47.400002       2014  
1774         Atlanta  48.200001       1897  
1773  Salt Lake City  49.070000       1971  
1772   New York City  49.380001       1899  
1771  Salt Lake City  50.000000       1910

### <font color='blue'>.sort_values</font>(<font color='gray'>'completed'</font><font color='red'>, ascending=False</font>) # biggest numbers first</font>

Alternatively, we can sort by numeric order of a column. The default way to do this is from lowest values to highest, but we can toggle the `ascending` argument to False to sort from highest values to lowest. 

In [42]:
sky_by_years = sky_df.sort_values('completed', ascending=False) # big numbers first
sky_by_years.head(8)

name         material               city  \
436             Amazon Tower I  mixed/composite            Seattle   
1768            City Hyde Park         concrete            Chicago   
582      Wolf Point West Tower         concrete            Chicago   
1025                   325 LEX         concrete      New York City   
1481                  Purves I         concrete      New York City   
1509                 One Light         concrete        Kansas City   
745   Chateau Beach Residences         concrete  Sunny Isles Beach   
1533    Rego Park Center Tower         concrete      New York City   

          height  completed  
436   159.720001       2016  
1768   52.389999       2016  
582   148.080002       2016  
1025  119.790001       2015  
1481   93.900002       2015  
1509   91.400002       2015  
745   136.800003       2015  
1533   89.919998       2015

### <font color='blue'>sort_values</font>(<font color='red'>['city', 'completed']</font>) # by data frame column list</font>

Sorting DataFrames is not limited to a single column. If we pass an array of columns to `.sort_values()`, we can sort each column in the order they're listed. In this case, the second column will be sorted nested within the first column. 

In [43]:
sky_by_city_year = sky_df.sort_values(['city', 'completed'])
sky_by_city_year.head(10)

name  material     city      height  \
1774                        FlatironCity     steel  Atlanta   48.200001   
775                       One Park Tower     steel  Atlanta  133.809998   
1075              230 Peachtree Building     steel  Atlanta  116.440002   
342            State of Georgia Building     steel  Atlanta  169.470001   
1140                  One Georgia Center     steel  Atlanta  113.080002   
1103   Coastal States Insurance Building     steel  Atlanta  114.900002   
103               Westin Peachtree Plaza  concrete  Atlanta  220.369995   
139                        AT&T Building  concrete  Atlanta  206.350006   
1519  Georgia Power Company Headquarters     steel  Atlanta   91.000000   
352               Marriott Marquis Hotel  concrete  Atlanta  168.860001   

      completed  
1774       1897  
775        1961  
1075       1965  
342        1967  
1140       1968  
1103       1971  
103        1976  
139        1980  
1519       1981  
352        1985

### <font color='blue'>sort_values(['city', 'completed'], </font><font color='red'>ascending=</font>[True, False]) 

We can also dictate the direction of our sort for each column passed. 

In [44]:
sky_by_city = sky_df.sort_values(['city', 'completed'], ascending=[True, False])
sky_by_city.head(15)

name         material     city  \
1647                         SkyHouse South         concrete  Atlanta   
575             1075 Peachtree Office Tower         concrete  Atlanta   
1049                          Loews Midtown         concrete  Atlanta   
289                            The Atlantic         concrete  Atlanta   
660                     3630 Peachtree Road         concrete  Atlanta   
946                          1010 Peachtree         concrete  Atlanta   
1009                    Two Alliance Center  mixed/composite  Atlanta   
1148                           Terminus 200         concrete  Atlanta   
1368  W Downtown Atlanta Hotel & Residences         concrete  Atlanta   
151                               Sovereign         concrete  Atlanta   
283                The Mansion on Peachtree         concrete  Atlanta   
522                               Viewpoint         concrete  Atlanta   
566          Twelve Centennial Park Tower I         concrete  Atlanta   
584                            Terminus 100         concrete  Atlanta   
156                          1180 Peachtree         concrete  Atlanta   

          height  completed  
1647   80.000000       2014  
575   148.740005       2010  
1049  118.190002       2010  
289   175.869995       2009  
660   142.949997       2009  
946   124.000000       2009  
1009  120.519997       2009  
1148  112.650002       2009  
1368  100.900002       2009  
151   202.690002       2008  
283   176.789993       2008  
522   152.699997       2008  
566   149.660004       2007  
584   147.910004       2007  
156   200.160004       2006

## Using DataFrame columns as 'numpy' arrays

Many vectorized operations (operations that apply iteratively to every element in an array) work specifically on 'numpy' arrays. This means its often important to extract columns in a DataFrame as an array object. This involves two steps: 

1) Identify the column   

2) Convert that column to numpy   


### my_array = df['column_name']<font color='blue'>.to_numpy()</font>

In [45]:
# Array from pandas data frame column
year_completed = sky_df['completed'].to_numpy()
year_completed

array([2014, 1974, 2015, ..., 1966, 1966, 2010], dtype=int64)

In [46]:
# Alternatively
np.array(sky_df['completed'])

array([2014, 1974, 2015, ..., 1966, 1966, 2010], dtype=int64)

### Compute a <font color='blue'>new age column</font>

Using the ability to perform vectorized operations on 'numpy' arrays, we can easily create a new column giving us the age of each building. 

In [47]:
sky_df['age'] = 2022 - sky_df['completed']
sky_df.head(5)

name         material           city  \
0             One World Trade Center  mixed/composite  New York City   
1                       Willis Tower            steel        Chicago   
2                    432 Park Avenue         concrete  New York City   
3  Trump International Hotel & Tower         concrete        Chicago   
4              Empire State Building            steel  New York City   

       height  completed  age  
0  541.299988       2014    8  
1  442.140015       1974   48  
2  425.500000       2015    7  
3  423.220001       2009   13  
4  381.000000       1931   91

## Grouping and reshaping data
<a id='Section4_2'></a>

### <font color='blue'>.groupby('column_name')</font>

The correlate to 'datascience' `.group()` is `.groupby()` in 'pandas. The application is basically the same: choose a column to group by, usually some sort of categorical variable or factor, and apply a method to each group. 

In [48]:
# Group by city and count the number of rows/observations associated with each
sky_by_city = sky_df.groupby('city')['city'].count()
sky_by_city.head(5)

city
Atlanta      38
Austin       17
Baltimore    14
Boston       32
Charlotte    12
Name: city, dtype: int64

In [49]:
# Group by city and find the max height (the tallest building) in each
sky_by_city = sky_df.groupby('city')['height'].max()
sky_by_city.head(6)

city
Atlanta      311.799988
Austin       208.149994
Baltimore    161.240005
Boston       240.789993
Charlotte    265.480011
Chicago      442.140015
Name: height, dtype: float64

In [50]:
# Group by city and compute the average age of tall buildings
sky_by_city_average = sky_df.groupby('city')['age'].mean()
sky_by_city_average.head(5)

city
Atlanta      28.710526
Austin       19.588235
Baltimore    44.571429
Boston       42.062500
Charlotte    27.250000
Name: age, dtype: float64

We're not limited to a single grouping factor - we can split up our data along any combination of variables. 

In [51]:
# Group by both city and material, then find the maximum height of each combination
sky_by_city_material = sky_df.groupby(['city', 'material'])['height'].max()
sky_by_city_material.head(7)

city       material       
Atlanta    concrete           264.250000
           mixed/composite    311.799988
           steel              169.470001
Austin     concrete           208.149994
           steel               93.599998
Baltimore  concrete           161.240005
           steel              155.149994
Name: height, dtype: float64

### <font color='blue'>.pivot_table()</font> # A core function in reshaping data, often better than groupby-apply sequences

This one is a long one, so we'll refer you to this guided tutorial:

The full URL
https://pbpython.com/pandas-pivot-table-explained.html 

Well worth checking it out. Starts with the basics, then adds a feature at a time to achieve some minor goal.

The excel file the examples use is in our github DSCI101 area so once you define path (in the usual way), you will be able to follow along with the examples.

In [52]:
# Change this 
# df = pd.read_excel("../in/sales-funnel.xlsx")
# df.head()

# To this -- then run the code shown in the Practical Business Python site (see URL above...)
df = pd.read_excel(path + 'sales-funnel.xlsx')
df.head()

# Enjoy the article, and enjoy the step by step approach

ImportError: Missing optional dependency 'openpyxl'.  Use pip or conda to install openpyxl.

: 

## Merge two DataFrames -- like datascience *join*
<a id='Section4_3'></a>

`.merge()` in 'pandas' (or `.join()` as you know it) takes two separate DataFrames and merges them into one using a **key**. A key is a common identifying variable shared between two datasets. Note that the presence of this key makes these datasets **relational**. 

### Read nba player data from 2013

In [ ]:
nba = pd.read_csv(path + 'nba2013.csv')
nba.sample(10)

### Read nba salary data

In [ ]:
nba_salary = pd.read_csv(path + 'nba_salaries.csv')
nba_salary.sample(10)

### <font color='blue'>merge left and right </font> data frames </font> -like <font color='gray'>datascience</font> <font color='darkgreen'>.join( )</font> 

When we talk about 'left' and 'right' DataFrames, we're typically talking about the original DataFrame to which we're adding new information (the left) and the DataFrame that is the source of this new information (the right). This is not always the case, merges can go in the other direction and sometimes we value information in both DataFrames equally, but is a useful convention. 

Let's demo two common merges.

### Left merge

The left merge is a merge that preserves all information in the left DataFrame, and incorporates only information from the right DataFrame with a matching key. Below we do a left merge using player name as our key (specifying the column name for our key in left and right DataFrames respectively). 

In [ ]:
nba_left = pd.merge(nba, nba_salary,                     
                   how='left',                         
                   left_on='Name',right_on='PLAYER')    
nba_left.head(9)

We end up with added information where Name (left) and PLAYER (right) are the same. Because we did a left merge, we preserve the entirety of the left DataFrame and therefore end up with null values where there was no match on our key. 

### Inner merge

The inner merge is also common, only keeping rows from both the left and right DataFrames where their key matches. Information from both DataFrames can be lost with this merge. 

In [ ]:
nba_all = pd.merge(nba, nba_salary,                     
                   how='inner',                         
                   left_on='Name',right_on='PLAYER')    
nba_all.head(9)

Other options include: 
- 'outer' merge where all information is kept from both DataFrames, resulting in null values added to both left and right DataFrames. 
- 'right' merge, where all information is preserved in the right DataFrame but not the left. 

## Remove missing values from your DataFrame

Removing null values can be important when calculating statistics or feeding data to models. For example, many functions that calculate averages will throw errors if fed missing values. 

### <font color='blue'>.dropna( )</font> drops all rows with any NaN value 

In [ ]:
nba_sal_plus = nba_all.dropna()

nba_sal_plus.sample(10)

## Replace values in your DataFrame

We can easily change values in our DataFrames with `.replace()`. This method is often invoked in data cleaning or feature engineering when values need tweaked, re-encoded, or corrected. 

In [ ]:
family = pd.read_csv(path + 'family_heights.csv')
family.head(3)

### Replace some values in a column

To change values in a column, we pass a dictionary to our method `.replace()` with `key:value` pairs corresponding to our old values on the left and new values on the right. 

In [ ]:
# Replace values in a column
family['sex'] = family['gender'].replace({'male': 0, 'female': 1}) # {old:new, old2:new2}
family.head(5)

### <font color='red'>'One hot encoding'</font> - changes category column into <font color='blue'>several 0/1 (yes/no) columns</font>, one column per unique category

Categorical variables (factors) can be both **ordinal**, meaning they have quantitative meaning *e.g. tax bracket*, and **nominal**, meaning they have no natural order or quantitative meaning *e.g. eye color*. **One hot encoding** is the process of turning a single nominal categorical variable into multiple binary (1/0) variables. This format is often more tractable in statistical models. 

In [ ]:
# Change column to "One Hot Encoding" using get_dummies

# In general, with n unique values, with get_dummies you will get n columns, all 0 except for rows
# that matche the column's category
#
sky_new = pd.get_dummies(sky_df, columns=['material'])
sky_new.head(4)

The 'material' column is now replaced with as many columns as unique values of material. Each value is either a 0, meaning not that material, or a 1, meaning yes that material. This process side-steps arbitrary ordering of our material variable - the factor levels initially corresponding to each variable were meaningless (*e.g.* concrete == 1, steel == 2, mix == 3) but still implied ordination. No such ordination is implied after one hot encoding. 

### Rename columns

In the same way we rename values, we can also rename columns by using a dictionary with the `.rename()` method. 

Handy when you need to use `df.query()`. Query must have columns with names that work as python variable names

In [ ]:
my_sky_df = sky_df.rename(columns={'name':'building', 'completed': 'year_built', })
my_sky_df.head()

# 5) Split data into Train and Test subsets
<a id='Section5'></a>

We often need to split our data into subsets, one for training some model and the other for testing the quality of our model. The following is a demonstration of one way to do this. 

In [ ]:
wine = pd.read_csv(path + 'wine.csv')
wine.info()

## To split into 89, 89 Train, Test, start with shuffle of all data

In [ ]:
# random_state= some number lets you rerun and get the same random data
#               not required, but handy for testing
shuffle_wine = wine.sample(frac=1.0, random_state=311)

In [ ]:
wine_train = shuffle_wine[0:89] # 1st 89 rows
wine_test = shuffle_wine[89:]   # remaining rows
wine_train.head(8)

In [ ]:
train_Class = wine_train['Class']
train_Class.head(5)

In [ ]:
train_attributes = wine_train.drop(columns=['Class'])
train_attributes.head(5)

In [ ]:
test_Actual_Class = wine_test['Class']
test_Actual_Class.head(5)

In [ ]:
test_attributes = wine_test.drop(columns=['Class'])
test_attributes.head(5)

# Quick reference 
<a id='Section6'></a>

In [ ]:
pandas handles tables of data; it calls a table a <font color='blue'>__DataFrame__</font> or df

```python 
import pandas as pd  
import matplotlib.pyplot as plt

# Read a csv file
my_df = pd.read_csv(path + 'my_file.csv') # instead of Table.read_table

my_df.head(6) # show 1st 6 rows of my_df data frame
my_df.tail(5) # show last 5 rows of my_df data frame

my_df.shape      # Shows number of rows, number of columns in your data frame
my_df.info()     # Shows each column name, number of non-empty items, type of column (int, float, object)
my_df.describe() # Shows count, mean, std, min, 25th percentile, 50th, 75th, max for each numeric column

# Rename some columns
my_df.rename(columns={'old col 1':'new_col1', 'another old': 'another_new', })

# variations on display of DataFrame rows
sky_sample_1_percent = sky_df.sample(frac=0.01) # random sample of 1% of rows in the data frame
sky_sample_6 = sky_df.sample(n=6) # random sample of 6 rows in the data frame
sky_sample_7 = sky_df.sample(n=7, random_state=311)   # random sample of 7, repeatable results

# .iloc[]  -- use index numbers for rows, columns to work with
sky_by_index = sky_df.iloc[3:6, :] # Rows 3, 4, 5 not 6, all columns -- like datascience .take()

my_df['a column'].unique() # Shows list of the unique value in this column

columns_list = ['name', 'age', 'major'] # Put columns in the order you want in your df
my_subset = my_df[columns_list] # Gets df with just those 3 columns  (like .select)

# sort_values
my_sorted_df = my_df.sort_values('major') # Sort by major, A to Z order
my_gpa_sort = sort_values('gpa', ascending=False) # biggest gpa numbers first

my_df2 = sort_values(['major', 'name']) # by major, within major by name
my_df3 = sort_values(['major', 'gpa'], ascending=[True, False]) # by major then by biggest gpa firs

# Convert df column to a numpy array  -- like datascience .column()
my_array = df['gpa'].to_numpy()

# Get selected rows from pandas data frame like datascience's .where( )

select_1979 = sky_df['completed'] == 1979
sky_1979_df = sky_df[select_1979] # Only put True rows into sky_1979_df 

select_before_1979 = sky_df['completed'] < 1979 
sky_before_1979_df = sky_df[select_before_1979] # Only put True rows into sky_before_1979_df 

# Get all the items that are in a list using .isin(list of values)
select_dates = sky_df['completed'].isin([1899, 1904, 1921]) # select dates is array of True / False values
sky_selected_years = sky_df[select_dates]

# Show buildings (built in 1924) & (taller than 150 meters) # Needs ( ) & ( )
select_1924_and_tall = (sky_df['completed'] == 1924) & (sky_df['height'] > 150.0)
sky_old_tall = sky_df[select_1924_and_tall]

# sky_old_tall = sky_df[select_1924 & select_tall] # another way
select_1924 = sky_df['completed'] == 1924
select_tall  = sky_df['height'] > 150.0
sky_old_tall3 = sky_df[select_1924 & select_tall]

# list of items from 1 column
select_cities = sky_df['city'].isin(['Atlanta', 'Sacramento', 'San Diego']) # array of True / False
sky_selected_cities = sky_df[select_cities]

# Either (completed in 1904 ) | (city is Sacramento)  # | means 'or'
select_city_or_date = (sky_df['city'] == 'Sacramento') | (sky_df['completed'] == 1904) 
sky_city_or_date = sky_df[select_city_or_date]

# df.query('city == "Chicago"') # works only when column names will be OK python variable names
sky_1980 = sky_df.query('completed == 1980')

# Create a new column
sky_df['age'] = 2021 - sky_df['completed'] # Stores result in named column on left of =

# Group by city, get counts
sky_by_city = sky_df.groupby('city')['city'].count()

# Group by city then within each city group, apply the max function to just the height column
sky_by_city = sky_df.groupby('city')['height'].max()

# Group by city and material, find tallest within each group
sky_by_city_material = sky_copy.groupby(['city', 'material'])['height'].max()

# pivot_table # The full URL https://pbpython.com/pandas-pivot-table-explained.html
df.pivot_table(index=["Manager","Rep"])

# merge -- like datascience join

# 3 varieties
# 1) inner merge (result must appear in both left and right data frames
nba_all = pd.merge(nba, nba_salary,                     # left is nba, right is nba_salary
                   how='inner',                         # must get data from both left and right data frames
                   left_on='Name',right_on='PLAYER')    # join on Name from left, PLAYER from right

# 2) left merge (result keeps every item in left data frame, matches with right when possible
#    Non-matches show as NaN ("Not a Number" -- missing value)
nba_players = pd.merge(nba, nba_salary,                 # left is nba, right is nba_salary
                   how='left',                          # gets all left data rows
                   left_on='Name',right_on='PLAYER')    # join on Name from left, PLAYER from right

# 3) right merge (result keeps every item in right data frame, matches with left when possible
#    Non-matches show as NaN ("Not a Number" -- missing value)
nba_salaries_plus = pd.merge(nba, nba_salary,            # left is nba, right is nba_salary
                    how='right',                         # gets all right data rows
                    left_on='Name',right_on='PLAYER')    # join on Name from left, PLAYER from right

# .dropna( )
nba_sal_plus = nba_salaries_plus.dropna( )  # .dropna() drops all rows with any NaN value

# Change values in your DataFrame
# Replace values in a column
family['gender'] = family['gender'].replace({'male': 0, 'female': 1}) # {old:new, old2:new2}

# 'One hot encoding' - changes category column into a series of yes/no columns
sky_new = pd.get_dummies(sky_df, columns=['material'])
# Material concrete, steel, 
# Kills material column, creates material_concrete, material_steel, 

# Options when reading .csv file

# keyword options with read_csv:
#  sep='\t' lets you read tab-separated data (or provide some other separator value)
#  header=None lets you read data without column names
#  names=['Who', 'What', 'Where'] provide your own column names
#  skiprows=3  Skips the first 3 rows (when you have notes and copyright before the real data)

# Shuffle data, then break into train, test data frames
shuffle_wine = wine.sample(frac=1.0, random_state=311) # Shuffle all data before a split
wine_train = shuffle_wine[0:89] # 1st 89 rows
wine_test = shuffle_wine[89:]   # remaining rows

```